tests/global_communication/test_communicator.cpp

```diff
@@ -14,22 +14,126 @@ using namespace nest::mc;
 
-    auto num_domains = policy::size();
-    auto rank = policy::id();
+    const auto num_domains = policy::size();
+    const auto rank = policy::id();
 
-    auto counts = policy.gather_all(1);
-    EXPECT_EQ(counts.size(), unsigned(num_domains));
-    for(auto i : counts) {
-        EXPECT_EQ(i, 1);
+    EXPECT_EQ(policy::min(rank), 0);
+    if (!is_dry_run()) {
+        EXPECT_EQ(policy::max(rank), num_domains-1);
     }
+}
+
+// some proxy types for communication testing
+struct source_proxy {
+    source_proxy() = default;
+    source_proxy(int g): gid(g) {}
+
+    int gid = 0;
+};
+
+bool operator==(int other, source_proxy s) {return s.gid==other;};
+bool operator==(source_proxy s, int other) {return s.gid==other;};
+
+struct spike_proxy {
+    spike_proxy() = default;
+    spike_proxy(int s, int d): source(s), domain(d) {}
+    source_proxy source = 0;
+    int domain = 0;
+};
+
+TEST(communicator, gather_spikes) {
+    using policy = communication::global_policy;
+
+    const auto num_domains = policy::size();
+    const auto rank = policy::id();
+
+    // dry run mode is a special case
+    if (is_dry_run()) {
+        const auto n_local_spikes = 10;
+        const auto n_local_cells = n_local_spikes;
 
-    auto part = util::parition_view(counts);
-    for(auto p : part) {
-        EXPECT_EQ(p.second-p.first, 1);
+        // Important: set up meta-data in dry run back end.
+        policy::set_sizes(policy::size(), n_local_cells);
+
+        // create local spikes for communication
+        std::vector<spike_proxy> local_spikes;
+        for (auto i=0; i<n_local_spikes; ++i) {
+            local_spikes.push_back(spike_proxy{i, rank});
+        }
+
+        // perform exchange
+        const auto global_spikes = policy::gather_spikes(local_spikes);
+
+        // test that partition information is correct
+        const auto& part = global_spikes.partition();
+        EXPECT_EQ(num_domains+1u, part.size());
+        for (auto i=0u; i<part.size(); ++i) {
+            EXPECT_EQ(part[i], n_local_spikes*i);
+        }
+
+        // test that spikes were correctly exchanged
+        //
+        // The local spikes had sources numbered 0:n_local_spikes-1
+        // The global exchange should replicate the local spikes and
+        // shift their sources to make them local to the "dummy" source
+        // domain.
+        // We set the model up with n_local_cells==n_local_spikes with
+        // one spike per local cell, so the result of the global exchange
+        // is a list of num_domains*n_local_spikes spikes that have
+        // contiguous source gid
+        const auto& spikes = global_spikes.values();
+        for (auto i=0u; i<spikes.size(); ++i) {
+            const auto s = spikes[i];
+            EXPECT_EQ(i, unsigned(s.source.gid));
+            EXPECT_EQ(0, s.domain);
+        }
+    }
+    else {
+        const auto scale = 10;
+        auto sumn = [scale](int n) {return scale*n*(n+1)/2;};
```

Comment:
Just randomly piping in: explicit is usually nicer because it makes the interdependencies clear. Especially if the use of the lambda leaves the scope in which it is defined.

Sometimes though a lambda is used in the same way that you would use any other code within a scope, and there e.g. `[&]` would make perfect sense. Depends on context as always.